napari/layers/tracks/_track_utils.py

```diff
@@ -0,0 +1,360 @@
+
+        # build a tree of the track data to allow fast lookup of nearest track
+        self._kdtree = cKDTree(self._points)
+
+        # make the lookup table
+        # NOTE(arl): it's important to convert the time index to an integer
+        # here to make sure that we align with the napari dims index which
+        # will be an integer - however, the time index does not necessarily
+        # need to be an int, and the shader will render correctly.
+        frames = list(set(self._points[:, 0].astype(np.uint).tolist()))
+        self._points_lookup = [None] * (max(frames) + 1)
+        for f in range(max(frames) + 1):
+            # if we have some data for this frame, calculate the slice required
+            if f in frames:
+                idx = np.where(self._points[:, 0] == f)[0]
+                self._points_lookup[f] = slice(min(idx), max(idx) + 1, 1)
+
+    @property
+    def properties(self) -> Dict[str, np.ndarray]:
+        """dict {str: np.ndarray (N,)}, DataFrame: properties for each track"""
+        return self._properties
+
+    @properties.setter
+    def properties(self, properties: Dict[str, np.ndarray]):
+        """ set track properties """
+        if not isinstance(properties, dict):
+            properties, _ = dataframe_to_properties(properties)
+
+        if 'track_id' not in properties:
+            properties['track_id'] = self.track_ids
+
+        # check the formatting of incoming properties data
+        self._properties = self._validate_track_properties(properties)
+
+    @property
+    def graph(self) -> Dict[str, list]:
+        """ dict {str: list}, graph representation of tracks """
+        return self._graph
+
+    @graph.setter
+    def graph(self, graph: Dict[str, list]):
+        """ set the track graph """
+
+        self._graph = self._validate_track_graph(graph)
+        # self.build_graph()
+
+    @property
+    def track_ids(self):
+        """ return the track identifiers"""
+        # return self.properties['track_id']
+        return self.data[:, 0].astype(np.uint16)
+
+    @property
+    def unique_track_ids(self):
+        """ return the unique track identifiers"""
+        return np.unique(self.track_ids)
+
+    def __len__(self):
+        """ return the number of tracks """
+        return len(self.unique_track_ids) if self.data is not None else 0
+
+    def _vertex_indices_from_id(self, track_id: int):
+        """ return the vertices corresponding to a track id """
+        return tuple(np.where(self.track_ids == track_id)[0])
+
+    def _validate_track_data(self, data: np.ndarray) -> np.ndarray:
+        """ validate the coordinate data """
+
+        if data.ndim != 2:
+            raise ValueError('track vertices should be a NxD array')
+
+        if data.shape[1] < 4 or data.shape[1] > 5:
+            raise ValueError('track vertices should be 4 or 5-dimensional')
+
+        if not all([idx.is_integer() for idx in data[:, 0].tolist()]):
```

Comment:
TIL about `npscalar.is_integer()`!!!

This seems a bit slow, perhaps we could do:

```suggestion
        # check that all IDs are integers
        ids = data[:, 0]
        if not np.all(np.floor(ids) == ids):
```
?